base/intfuncs.jl

```diff
@@ -82,14 +82,16 @@ end
 
 gcd(a::Union{Integer,Rational}) = a
 lcm(a::Union{Integer,Rational}) = a
-gcd(a::Union{Integer,Rational}, b::Union{Integer,Rational}) = gcd(promote(a,b)...)
-lcm(a::Union{Integer,Rational}, b::Union{Integer,Rational}) = lcm(promote(a,b)...)
-gcd(a::Union{Integer,Rational}, b::Union{Integer,Rational}...) = gcd(a, gcd(b...))
-lcm(a::Union{Integer,Rational}, b::Union{Integer,Rational}...) = lcm(a, lcm(b...))
+gcd(a::Real, b::Real) = gcd(promote(a,b)...)
+lcm(a::Real, b::Real) = lcm(promote(a,b)...)
+gcd(a::Real, b::Real...) = gcd(a, gcd(b...))
+lcm(a::Real, b::Real...) = lcm(a, lcm(b...))
+gcd(a::T, b::T) where T<:Real = throw(MethodError(gcd, (a,b)))
```

Comment:
Manually throwing method errors tend to often be a code smell from my experience. What is the reason for it? Is it to avoid an infinite recursion?

Reply:
Yes, the reason is to avoid infinite recursion, since some `Real` types don’t implement `gcd`/`lcm`/`gcdx` (`AbstractFloat`s, for example).